draft-ietf-mls-protocol.md

```diff
@@ -1959,54 +1956,36 @@ leaf uses the j-th key/nonce pair in the sequence to encrypt (using the AEAD)
 the j-th message they send during that epoch. In particular, each key/nonce pair
 MUST NOT be used to encrypt more than one message.
 
-More precisely, the initial secret of the ratchet for the group member assigned to the leaf with index i is simply the secret of that leaf.
-
-~~~~
-application_[i]_[0]_secret = astree_node_[i]_secret
-~~~~
-
-Keys, nonces and secrets of ratchets are derived using HKDF-Expand-Label. The
-context in a given call consists of (a hash of) the current Group state, the
-index of the sender's leaf in the ratchet tree and the position in the ratchet.
-In particular, the index of the sender's leaf in the ratchet tree is the same
-as the index of the leaf in the AS Tree used to initialize the sender's ratchet.
+More precisely, the initial secret of the ratchet for the group
+member assigned to the leaf with node index V is simply the secret of
+that leaf.
 
 ~~~~
-struct {
-  opaque gshash<0..2^32-1> = Hash(GroupState_[n]);
-  uint32 leaf_index;
-  uint32 ratchet_position;
-} RatchetContext
+application_[V]_[0]_secret = astree_node_[V]_secret
 ~~~~
 
-Ratchet_position is initialized to 1 for the first HKDF-Expand-Label call of a
-new ratchet and incremented for each subsequent call in that ratchet.
-
-HashRatCont[i, j] denotes a RatchetContext variable where:
-
-- gshash = Hash(GropuState_[n])
-
-- leaf_index = i
-
-- ratchet_position = j
-
-Using this notation the ratchet for the group member at leaf i is defined to be:
+Keys, nonces and secrets of ratchets are derived using
+Derive-App-Secret. The context in a given call consists of the index
+of the sender's leaf in the ratchet tree and the current position in
+the ratchet.  In particular, the index of the sender's leaf in the
+ratchet tree is the same as the index of the leaf in the AS Tree
+used to initialize the sender's ratchet.
 
 ~~~~
-application_[i]_[j]_secret
+application_[V]_[j]_secret
       |
-      +--> HKDF-Expand-Label(., "app-nonce", HashRatCont[i,j+1], AEAD.nonceLen)
-      |    = application_[i]_[j+1]_nonce
+      +--> Derive-App-Secret(., "app-nonce", V, j, AEAD.nonce_length)
```

Comment:
This is not an App-Secret derivation here, I would prefer to stick to a standard HKDF-Expand-Label for these.

Reply:
But I agree that it make sense for the app-secret to app-secret derivation below.

Reply:
Do you think this is harmful, or just un-aesthetic?  The goal here is to preserve Joël's invariant that every derivation has different context.